runtime/src/accounts_index.rs

```diff
@@ -181,9 +211,201 @@ impl<'a, T: 'static + Clone> Iterator for AccountsIndexIterator<'a, T> {
+                    .downgrade()
+            });
+            let should_insert = {
+                // Most of the time, key should already exist and match
+                // the one in the update
+                if let Some(existing_key) = slots_map.read().unwrap().get(&slot) {
+                    existing_key != key
+                } else {
+                    // If there is no key yet, then insert
+                    true
+                }
+            };
+            if should_insert {
+                slots_map.write().unwrap().insert(slot, *key)
+            } else {
+                None
+            }
+        };
+
+        if let Some(prev_key) = prev_key {
+            // If the inner key was moved to a different primary key, remove
+            // the previous index entry.
+            if prev_key != *key {
+                self.remove_index_entries(&prev_key, inner_key, &[slot]);
+            }
+        }
+    }
+
+    fn remove_index_entries(&self, key: &Pubkey, inner_key: &Pubkey, slots: &[Slot]) {
+        let is_key_empty = if let Some(inner_key_map) = self.index.get(&key) {
+            // Delete the slot from the slot set
+            let is_inner_key_empty = if let Some(slot_set) = inner_key_map.get(&inner_key) {
+                let mut w_slot_set = slot_set.write().unwrap();
+                for slot in slots.iter() {
+                    let is_present = w_slot_set.remove(slot);
+                    if !is_present {
+                        warn!("Reverse index is missing previous entry for key {}, inner_key: {}, slot: {}",
+                        key, inner_key, slot);
+                    }
+                }
+                w_slot_set.is_empty()
+            } else {
+                false
+            };
+
+            // Check if `key` is empty
+            let is_key_empty = if is_inner_key_empty {
+                if let dashmap::mapref::entry::Entry::Occupied(inner_key_entry) =
+                    inner_key_map.entry(*inner_key)
+                {
+                    // Delete the `inner_key` if the slot set is empty
+                    let slot_set = inner_key_entry.get();
+
+                    // Write lock on `inner_key_entry` above through the `entry`
+                    // means nobody else has access to this lock at this time,
+                    // so this check for empty -> remove() is atomic
+                    if slot_set.read().unwrap().is_empty() {
+                        inner_key_entry.remove();
+                    }
+                }
+                inner_key_map.is_empty()
+            } else {
+                false
+            };
+            is_key_empty
+        } else {
+            false
+        };
+
+        // Delete the `key` if the set of inner keys is empty
+        if is_key_empty {
+            if let dashmap::mapref::entry::Entry::Occupied(key_entry) = self.index.entry(*key) {
+                if key_entry.get().is_empty() {
```

Comment:
maybe add comment like this considering your comment(https://github.com/solana-labs/solana/pull/14212/files#r547081588)?:

```
// Other threads might interleaved by having written to this `key`,
// so double-check again for its emptiness 
```

Reply:
agreed, more documentation the better in this code path!